src/foam/nanos/notification/email/SMTPEmailService.js

```diff
@@ -190,27 +193,45 @@ foam.CLASS({
       `
     },
     {
-      name: 'sendEmail',
+      name: 'execute',
       javaCode:
       `
-      OMLogger omLogger = (OMLogger) x.get("OMLogger");
-      Logger logger = (Logger) getX().get("logger");
-        try {
-          omLogger.log("Pre send email request");
-          MimeMessage message = createMimeMessage(emailMessage);
-          Transport transport = getSession_().getTransport("smtp");
-          transport.connect();
-          logger.info("SMTPEmailService connected.");
-          transport.send(message, getUsername(), getPassword());
-          logger.info("SMTPEmailService sent MimeMessage.");
-          transport.close();
-          logger.info("SMTPEmailService finish.");
-          omLogger.log("Post send email request");
-        } catch (Exception e) {
-          logger.error("SMTPEmailService failed to finish. " + e);
-        }
+        DAO emailMessageDAO = (DAO) x.get("emailMessageDAO");
+        List<EmailMessage> emailMessages = ((foam.dao.ArraySink)
+          emailMessageDAO
+            .where(EQ(EmailMessage.STATUS, Status.UNSENT))
+            .select(new ArraySink()))
+            .getArray();
         
-      `
+            OMLogger omLogger = (OMLogger) x.get("OMLogger");
+            Logger logger = (Logger) getX().get("logger");
+        
+            try {
+              omLogger.log("Pre send email request");
+              Transport transport = getSession_().getTransport("smtp");
+              transport.connect();
+              logger.info("SMTPEmailService connected.");
+              for (EmailMessage emailMessage : emailMessages) {
+                MimeMessage message = createMimeMessage(emailMessage);
+                try {
+                  transport.send(message, getUsername(), getPassword());
+                  emailMessage.setStatus(Status.SENT);
+                  logger.info("SMTPEmailService sent MimeMessage.");
+                } catch ( MessagingException e ) {
+                  emailMessage.setStatus(Status.FAILED);
+                  logger.error("SMTPEmailService sending MimeMessage failed. " + e);
+                } finally {
+                  emailMessageDAO.put(emailMessage);
+                }
+              }
+              transport.close();
```

Comment:
put transport close in a finally block

Reply:
I think it's better to leave it here or in the try block because:
1.
https://stackoverflow.com/questions/4334027/is-it-acceptable-to-leave-the-javamail-session-transport-open
2.
transport.close() itself throws exception